packaging.md

```diff
@@ -0,0 +1,156 @@
+
+See [Docker's documentation](https://docs.docker.com/engine/reference/commandline/) for more complete information about Docker.
+
+### Distroless
+
+[Distroless](https://github.com/GoogleContainerTools/distroless) is a project by Google that attempts to create minimal images containing only the application and its runtime dependencies. They do not contain package managers, shells or any other programs you would expect to find in a standard Linux distribution.
+
+Since distroless supports Docker and is based on Debian, packaging a Swift application on it is fairly similar to the Docker process above. Here is an example `Dockerfile` that builds and packages the application on top of a distroless's C++ base image:
+
+```Dockerfile
+#------- build -------
+# Building using Ubuntu Bionic since its compatible with Debian runtime
+FROM swiftlang/swift:nightly-bionic as builder
+
+# set up the workspace
+RUN mkdir /workspace
+WORKDIR /workspace
+
+# copy the source to the docker image
+COPY . /workspace
+
+RUN swift build -c release
+
+#------- package -------
+# Running on distroless C++ since it includes
+# all(*) the runtime dependencies Swift programs need
+FROM gcr.io/distroless/cc-debian10
+# copy executables
+COPY --from=builder /workspace/.build/release/<executable-name> /
+# copy Swift's dynamic libraries dependencies
+COPY --from=builder /usr/lib/swift/linux/lib*so* /
+
+# set the entry point (application name)
+CMD ["<executable-name>"]
+```
+
+Note the above uses `gcr.io/distroless/cc-debian10` as the runtime image which should work for Swift programs that do not use `FoundationNetworking` or `FoundationXML`. In order to provide more complete support we (the community) could put in a PR into distroless to introduce a base image for Swift that includes `libcurl` and `libxml` which are required for `FoundationNetworking` and `FoundationXML` respectively.
+
+## Archive (Tarball, ZIP file, etc.)
+
+Since cross-compiling Swift for Linux is not (yet) supported on Mac or Windows, we need to use virtualization technologies like Docker to compile applications we are targeting to run on Linux.
+
+That said, this does not mean we must also package the applications as Docker images in order to deploy them. While using Docker images for deployment is convenient and popular, an application can also be packaged using a simple and lightweight archive format like tarball or ZIP file, then uploaded to the server where it can be extracted and run.
+
+Here is an example of using Docker and `tar` to build and package the application for deployment on Ubuntu servers:
+
+First, use the `docker run` command from the application's source location to build it:
+
+```bash
+$ docker run --rm \
+  -v "$PWD:/workspace" \
+  -w /workspace \
+  swift:5.2-bionic \
+  /bin/bash -cl "swift build -c release"
```

Comment:
```suggestion
  /bin/bash -xcl "swift -version && swift build -c release"
```

feel free to ignore but I prefer to always see the version number as that's often important.

Reply:
I also like `-x` because it prints all the random env variables that may be important when debugging stuff (but it does make the output a little bit uglier ofc)

Reply:
sample output:

```
$ jw-docker-swift-5.2 /bin/bash -xcl "swift -version && swift build -c release"
+ '[' '' ']'
+ '[' -d /etc/profile.d ']'
+ for i in /etc/profile.d/*.sh
+ '[' -r /etc/profile.d/01-locale-fix.sh ']'
+ . /etc/profile.d/01-locale-fix.sh
+++ /usr/bin/locale-check C.UTF-8
++ eval
+ unset i
+ '[' /bin/bash ']'
+ '[' -f /root/.bashrc ']'
+ . /root/.bashrc
++ '[' -z '' ']'
++ return
+ mesg n
+ true
+ swift -version
Swift version 5.2.3 (swift-5.2.3-RELEASE)
Target: x86_64-unknown-linux-gnu
+ swift build -c release
[39/39] Linking NIOPerformanceTester
```